### PR TITLE
chore: migrate src/Page.js to TypeScript

### PR DESCRIPTION
**Pull request by jackfranklin**
The final one! This is a huge file and needs to be split up and tidied,
but for now I've left all the definitions in place and converted types
accordingly.

There's some additional tidying we can do now every `src` file is TS,
but I'll leave that for another PR to avoid this one getting any bigger.

### CODE REVIEW

File: src/Dialog.ts [hunk 1]
```diff
@@ -17,14 +17,17 @@
 import {assert} from './helper';
 import {CDPSession} from './Connection';
 
-enum DialogType {
+/* TODO(jacktfranklin): protocol.d.ts defines this
```
Comment (jackfranklin) [inline src/Dialog.ts:20]: follow up PR

File: src/externs.d.ts [hunk 2]
```diff
@@ -1,14 +1,7 @@
...
   module Puppeteer {
-    export class Page extends RealPage { }
-    /* TODO(jacktfranklin@): once DOMWorld, Page, and FrameManager are in TS
-     * we can remove this and instead use the type defined in LifeCycleWatcher
-     */
-    export type PuppeteerLifeCycleEvent = 'load' | 'domcontentloaded' | 'networkidle0' | 'networkidle2';
-
     export interface ConnectionTransport {
       send(string);
       close();
```
Comment (jackfranklin) [inline src/externs.d.ts:7]: in a follow up PR I will remove this file and find a new home for the interfaces we need where they are not exposed globally

File: utils/doclint/check_public_api/index.js [hunk 3]
```diff
@@ -359,6 +359,46 @@ function compareDocumentations(actual, expected) {
...
+        expectedName: 'ScreenshotOptions'
+      }],
+      ['Method Page.setContent() options', {
+        actualName: 'Object',
+        expectedName: 'WaitForOptions'
+      }],
+      ['Method Page.setCookie() ...cookies', {
+        actualName: '...Object',
+        expectedName: '...CookieParam'
+      }],
```
Comment (jackfranklin) [inline utils/doclint/check_public_api/index.js:401]: now we're in TS land we can start diving properly into documentation tools!

File: src/Page.ts [hunk 4]
```diff
@@ -14,92 +14,160 @@
...
+  _frameManager: FrameManager;
+  _emulationManager: EmulationManager;
+  _tracing: Tracing;
+  _pageBindings = new Map<string, Function>();
+  _coverage: Coverage;
+  _javascriptEnabled = true;
+  _viewport?: Puppeteer.Viewport;
+  _screenshotTaskQueue: TaskQueue;
+  _workers = new Map<string, PuppeteerWorker>();
+  // TODO: improve this typedef - it's a function that takes a file chooser or something?
```
Comment (jackfranklin) [inline src/Page.ts:152]: will do in a follow up PR

File: src/Dialog.ts (hunk 1 above)
Comment (mathiasbynens) [reply src/Dialog.ts:20]: ack

File: utils/doclint/check_public_api/JSBuilder.js [hunk 5]
```diff
@@ -204,6 +204,7 @@ function checkSources(sources) {
 
     for (const [name, member] of symbol.members || []) {
 
+
```
Comment (mathiasbynens) [inline utils/doclint/check_public_api/JSBuilder.js]: This seems accidental?

```suggestion
```

File: utils/doclint/check_public_api/index.js (hunk 3 above)
Comment (mathiasbynens) [reply utils/doclint/check_public_api/index.js:401]: \o/ so good!

**Review by mathiasbynens — APPROVED** [review]
lgtm

**Pull request merged**